### PR TITLE
Migrate migration log to have predictable id

### PR DESCRIPTION
**Pull request by garethbowen**
Make it easier to find the migration log as it turns out it's useful
information for humans as well as machines.

medic/medic-webapp#3294

### CODE REVIEW

**Comment by alxndrsn** [comment]
I don't see an accompanying commit in `medic-webapp`; is this doc only used in `medic-api`?

**Comment by garethbowen** [comment]
@alxndrsn Yeah it's only used for recording which migrations have been executed.

File: migrations.js [hunk 1]
```diff
@@ -3,33 +3,75 @@ var fs = require('fs'),
...
+
+var deleteOldLog = function(oldLog, callback) {
+  if (!oldLog) {
+    return callback();
+  }
+  oldLog._deleted = true;
+  db.medic.insert(oldLog, callback);
+};
+
+var createLog = function(callback) {
```
Comment (alxndrsn) [inline migrations.js]: Not convinced by this name - this is the function that migrates from old log doc to the new one IIUC, so perhaps `migrateLogToKnownId()` or something would be more appropriate?

File: migrations.js [hunk 2]
```diff
@@ -79,15 +121,18 @@ var runMigrations = function(meta, migrations, callback) {
 
 module.exports = {
   run: function(callback) {
-    getMeta(function(err, meta) {
+    module.exports.get(function(err, migrations) {
```
Comment (alxndrsn) [inline migrations.js]: Appreciate this is not part of this change, but is `get()` only exported for unit tests?

File: migrations.js [hunk 3]
```diff
@@ -79,15 +121,18 @@ var runMigrations = function(meta, migrations, callback) {
...
       if (err) {
         return callback(err);
       }
-      module.exports.get(function(err, migrations) {
+      if (!migrations.length) {
```
Comment (alxndrsn) [inline migrations.js]: Is this an alternative to checking `if(err)`?  Can't see why the migrations list would ever be falsy, otherwise.

File: migrations.js [hunk 4]
```diff
@@ -3,33 +3,75 @@ var fs = require('fs'),
...
+
+var getLog = function(callback) {
+  db.medic.get(MIGRATION_LOG_ID, function(err, doc) {
+    if (!err) {
+      return callback(null, doc);
     }
-    if (!meta.migrations) {
-      meta.migrations = [];
+    if (err.statusCode === 404) {
+      return createLog(callback);
```
Comment (alxndrsn) [inline migrations.js]: Doesn't look like `createLog()` would ever return anything.  Is this return just to shortcut out of this function, or is there meant to be a chained promise here?

File: migrations.js (hunk 1 above)
Comment (garethbowen) [inline migrations.js]: It migrates the old log if it's there (upgrade), or creates an empty one if it's not there (fresh install). I've renamed it to `createMigrationLog` because it's creating the new doc style ont way or another.

File: migrations.js (hunk 4 above)
Comment (garethbowen) [inline migrations.js]: There are no promises in this code anywhere, mostly because we still use nano which is callback based. One day...

I've cleaned up the code a little to hopefully make it easier to read.

File: migrations.js (hunk 2 above)
Comment (garethbowen) [inline migrations.js]: Yeah I think that's right. IIRC when writing this originally it was too hard to mock out reading the migrations dir.

File: migrations.js (hunk 3 above)
Comment (garethbowen) [inline migrations.js]: That's a good point. We have hardcoded at least one migration! I got a bit overzealous with edge cases. I'll remove it.
